server/chat-plugins/chatlog.ts

```diff
@@ -319,6 +440,121 @@ const LogViewer = new class {
+
+	render(results: string[], roomid: RoomID, search: string, cap?: number) {
+		const matches = [];
+		let curDate = '';
+		const dates = [];
+		for (const chunk of results.sort()) {
+			const rebuilt: string[] = [];
+			const exacts = [];
+			exacts.push(chunk.split('\n').filter((item: string) => new RegExp(search, "i").test(item)).map(item => {
+				item = item.split(item.includes('.txt-') ? '.txt-' : '.txt:')[1];
+				item = LogViewer.renderLine(item);
+				return item;
+			})[0]); // get exact match for display
+
+			for (const line of chunk.split('\n')) {
+				const sep = line.includes('.txt-') ? '.txt-' : '.txt:';
+				const [name, text] = line.split(sep);
+				let rendered = LogViewer.renderLine(text, 'all');
+				if (!rendered || name.includes('today') || !toID(line)) continue;
+				 // gets rid of some edge cases / duplicates
+				let date = name.replace(`${__dirname}/../../logs/chat/${roomid}`, '').slice(9);
+				if (curDate !== date) {
+					curDate = date;
+					dates.push(curDate);
+					date = `</details><details><summary>[<a href="view-chatlog-${roomid}--${date}">${date}</a>]</summary>`;
+					rendered = `${date} ${rendered}`;
+				} else {
+					date = '';
+				}
+				const matched = (
+					new RegExp(search, "i")
+						.test(rendered) ? `<div class="chat chatmessage highlighted">${rendered}</div>` : rendered
+				);
+				rebuilt.push(matched);
+			}
+			const isIn = matches.join(' ').includes(rebuilt.join(' '));
+			if (cap && matches.push(rebuilt.join(' ')) >= cap) {
+				break;
+			} else {
+				// `isIn` resolves a duplication bug
+				if (isIn) matches.push(rebuilt.join(' '));
+			}
+		}
+		let buf = `<div class="pad"><p><strong>Results on ${roomid} for ${search}:</strong>`;
+		let total = 0;
+		for (const match of matches.join(' ').split(' ')) {
+			if (new RegExp(search, "i").test(match)) total++;
+		}
+		buf += ` ${total}`;
+		if (cap) {
+			buf += ` (capped at ${cap})<hr/ >`;
+		} else {
+			buf += `<hr/ >`;
+		}
+		buf += matches.join('<hr/ >');
+		if (cap) {
+			buf += `<hr/ ><strong>Capped at ${cap}.</strong><br>`;
+			buf += `<button class="button" name="send" value="/sl ${search}, ${roomid},,${cap + 200}">View 200 more<br />&#x25bc;</button>`;
+			buf += `<button class="button" name="send" value="/sl ${search},${roomid},,all">View all<br />&#x25bc;</button></div>`;
+		}
+		return buf;
+	}
+
+	async ripgrepSearch(roomid: RoomID, search: string, cap: number) {
+		let output;
+		try {
+			const options = [
+				search,
+				`${__dirname}/../../logs/chat/${roomid}`,
+				'-C', '3',
+			];
+			output = await execFile('rg', options, {cwd: path.normalize(`${__dirname}/../../`)});
+		} catch (error) {
+			return LogViewer.error(
+				`No results found.`
+			);
+		}
+		const matches = [];
+		for (const result of output.stdout.split('--').reverse()) {
+			matches.push(result);
+		}
+		return this.render(matches, roomid, search, cap);
+	}
+
+	async grepSearch(roomid: RoomID, search: string, cap?: number): Promise<string> {
```

Comment:
What's the point of still supporting grep?

Reply:
I figured i'd leave it for side servers that wanted a faster option.

Reply:
Ripgrep is the faster option not grep.

Reply:
Fair. I was thinking for people who don't want to / can't install it, but that's a decent point.